src/query_graph/output.rs

```diff
@@ -0,0 +1,149 @@
+    if graph.sources.len() > 1 {
+        return to_dot_federated(graph).expect("Failed to generate the federated graph");
+    }
+
+    // Note: Use label_edge/label_node as `attr_getters` in order to create custom label
+    //       strings, instead of the default labeling.
+    let config = [Config::NodeNoLabel, Config::EdgeNoLabel];
+    Dot::with_attr_getters(
+        &graph.graph,
+        &config,
+        &(|_, er| label_edge(er.weight())),
+        &(|_, (_, node)| label_node(node)),
+    )
+    .to_string()
+}
+
+fn to_dot_federated(graph: &QueryGraph) -> Result<String, std::fmt::Error> {
+    fn edge_within_cluster(
+        graph: &StableInnerGraph,
+        cluster_name: &NodeStr,
+        edge_index: EdgeIndex,
+    ) -> bool {
+        match graph.edge_endpoints(edge_index) {
+            None => false,
+
+            Some((n1, n2)) => {
+                graph[n1].source == *cluster_name && graph[n2].source == *cluster_name
+            }
+        }
+    }
+
+    fn edge_across_clusters(graph: &StableInnerGraph, edge_index: EdgeIndex) -> bool {
+        match graph.edge_endpoints(edge_index) {
+            None => false,
+
+            Some((n1, n2)) => graph[n1].source != graph[n2].source,
+        }
+    }
+
+    fn label_cluster_node(node: &QueryGraphNode) -> String {
+        format!(r#"label="{}@{}""#, node.type_, node.source)
+    }
+
+    // Build a stable graph, so we can derive subgraph clusters with the same indices.
+    let stable_graph = StableGraph::from(graph.graph.clone());
+    let cluster_dot_config = [
+        Config::NodeNoLabel,
+        Config::EdgeNoLabel,
+        Config::GraphContentOnly,
+    ];
+
+    let mut dot_str = String::new();
+    writeln!(dot_str, r#"digraph "{}" {{"#, graph.name())?;
+
+    // Subgraph clusters
+    for (cluster_name, _) in graph.sources.iter() {
+        if cluster_name == graph.name() {
+            continue; // skip non-subgraph nodes
+        }
+
+        let filtered_graph: StableInnerGraph = stable_graph.filter_map(
+            |_i, n| {
+                if n.source == *cluster_name {
+                    Some(n.clone())
+                } else {
+                    None
+                }
+            },
+            |i, e| {
+                if edge_within_cluster(&stable_graph, cluster_name, i) {
+                    Some(e.clone())
+                } else {
+                    None
+                }
+            },
+        );
+        let s = Dot::with_attr_getters(
+            &filtered_graph,
+            &cluster_dot_config,
+            &(|_, er| label_edge(er.weight())),
+            &(|_, (_, node)| label_cluster_node(node)),
+        )
+        .to_string();
+
+        writeln!(dot_str, r#"  subgraph "cluster_{}" {{"#, &cluster_name)?;
+        writeln!(dot_str, r#"    label = "Subgraph \"{}\"";"#, &cluster_name)?;
+        writeln!(dot_str, r#"    color = "black";"#)?;
+        writeln!(dot_str, r#"    style = """#)?;
```

Comment:
I'm not too familiar with dot syntax, but does this need a semicolon?
```suggestion
        writeln!(dot_str, r#"    style = "";"#)?;
```

Reply:
Ah, it looks like it's optional.

Reply:
I believe so. I tried mimic the JS implementation's behavior :)